docs/latest/design/ratelimit.md

```diff
@@ -0,0 +1,176 @@
+# Rate Limit Design
+
+## Overview
+
+Rate limit is a feature that allows the user to limit the number of incoming requests
+to a predefined value based on attributes within the traffic flow.
+
+Here are some reasons why a user may want to implements Rate limits
+
+* To prevent malicious activity such as DDoS attacks.
+* To prevent applications and its resources (such as a database) from getting overloaded.
+* To create API limits based on user entitlements.
+
+## Scope Types
+
+The rate limit type here describe the scope of rate limits
+
+* Global - In this case, the rate limit is common across all the instances of Envoy proxies
+where its applied i.e. if the data plane has 2 replicas of Envoy running, and the rate limit is
+10 requests/second, this limit is common and will be hit if 5 requests pass through the first replica
+and 5 requests pass through the second replica within the same second.
+
+* Local - In this case, the rate limits are specific to each instance/replica of Envoy running.
+Note - This is not part of the initial design and will be added as a future enhancement. 
+
+## Match Types 
+
+### Rate limit a specific traffic flow 
+
+* Here is an example of a ratelimit implemented by the application developer to limit a specific user
+by matching on a custom `x-user-id` header with a value set to `one`
+
+```yaml
+apiVersion: gateway.envoyproxy.io/v1alpha1
+kind: RateLimitFilter
+metadata:
+  name: ratelimit-specific-user
+spec:
+  type: Global
+  rules:
+  - matches:
```

Comment:
Since the RateLimitFilter API is being used as a [filter](https://gateway-api.sigs.k8s.io/references/spec/#gateway.networking.k8s.io/v1beta1.HTTPRouteFilter), I don;t understand what `rules` and `matches` is needed. An HTTProute already has rules and matches. I would expect a RateLimitFilter of type `Global` to define descriptors that are sent to the rate limit service.

Reply:
imho descriptors are an underlying implementation that shouldnt be exposed to app developers

Reply:
I'm fine not using `descriptors`, we should define a global rate-limiting schema that provides flexibility for defining rate limits. IMO, Contour's [GlobalRateLimitPolicy](https://github.com/projectcontour/contour/blob/v1.23.2/apis/projectcontour/v1/httpproxy.go#L665-L673) is a good reference for this effort.

@arkodg aside from [envoyproxy/ratelimit](https://github.com/envoyproxy/ratelimit), what other global rate-limiting solutions exist that we should consider supporting? This will help guide the development of the rate-limiting API. If we expect to support different global rate-limiting implementations, then either the API should abstract implementation-specific APIs or the RateLimitingFilter directly exposes the implementation-specific APIs (similar to our plans for having EG expose the xDS APIs directly). For example:
```
apiVersion: gateway.envoyproxy.io/v1alpha1
kind: RateLimitFilter
metadata:
  name: ratelimit-specific-user
spec:
  # Only 1 of type "EnvoyGlobal" or "Local" may be specified.
  # "EnvoyGlobal" uses https://github.com/envoyproxy/ratelimit as the RL service.
  # Other global rate limit implementations
  type: EnvoyGlobal
  envoyGlobal:
    domain:
    descriptors:
    # All fields of envoyGlobal are specific to the envoyproxy/ratelimit implementation.
...
```

`EnvoyProxy
cc: @skriss @sunjayBhatia @youngnick

Reply:
With ^ approach, we could state that RateLimitFilter only supports Envoy for local and envoyproxy/ratelimit for global. If other implementations that integrate with Envoy Proxy are interesting in adding rate limiting support, they would submit a PR to add their API and implement said API. For example:
```
apiVersion: gateway.envoyproxy.io/v1alpha1
kind: FooRateLimitFilter
metadata:
  name: ratelimit-specific-user
spec:
  ...
```

Reply:
I think that actually writing out the example for doing multiple overlapping rate limits on a single HTTPRoute might help clarify this. 🙂

Reply:
the descriptor approach we took in contour isn't intended to be specific to the envoy/ratelimit project, but should be applicable to any implementation that supports the RLS API: https://www.envoyproxy.io/docs/envoy/latest/api-v3/service/ratelimit/v3/rls.proto, which is the only way Envoy has to configure a remote ratelimit server

Reply:
the `RateLimitFilter` design is intentional in leaving out any Envoyisms out of it. This is in line with the [goals ](https://github.com/envoyproxy/gateway/blob/main/GOALS.md)of the project to provide a simple API for application devlopers (`Application developers will use the API out of the box without needing to understand in-depth concepts of Envoy Proxy or use OSS wrappers`)

For users familiar with Envoy APIs and wanting more than the presented API, should be able to do so with the full power of [xds ](https://github.com/envoyproxy/gateway/blob/main/GOALS.md#extensibility) which is being tracked with https://github.com/envoyproxy/gateway/issues/24

Reply:
might be good to take a step back and describe what needs to be provided to users, as I see it:
- a way to specify what the *rate* part of the rate limit is
- what subset of traffic it applies to (hostnames, routes, using matches etc.)
- how that request traffic is communicated to the RLS, i.e. how the RLS differentiates requests to rate limit

The third part you kinda need knowledge of how Envoys API works and is where I think the debate over matches/descriptors etc. comes in

one option would be to not provide any sub-matching/descriptor support initially and see what needs people have over time, so a nice abstraction can be grow more from user feedback, not sure if we've really gotten that yet for this feature or are trying to match Contour/assumptions about what people need

Reply:
I kinda feel like we're going in several directions at once here. 🙂 Let me see if I can summarize the whole thing...

Fundamentally, the hard part of ratelimiting isn't having a mechanism by which Envoy can return a 429 to a request: that could be the existing Envoy ratelimit filter, or a WASM or Lua filter, or a custom filter... even the extauthz filter has that power. The hard part - and @danehans and @sunjayBhatia are (I believe?) getting at this - is giving the user a coherent, explainable set of concepts that they can use to give that decision-making mechanism enough information to make reasonable decisions.

In current Envoy, that set of concepts is descriptors, so whatever we do in Envoy Gateway needs to present something to the user that either

a. **is** descriptors, or
b. ultimately is mappable - in an explainable, consistent way - back to descriptors(*).

Given that descriptors are actually kind of unpleasant to work with in their current form, I would love to have them **not** be the mechanism we expose to the user. In that case, though, we **absolutely** need an in-depth, detailed discussion of how the concept being proposed maps back to Envoy descriptors. That could be in the `RateLimitFilter` design document (I personally think it should be) or in a separate document, but it must be written down, and it needs to take the multiple-overlapping-ratelimits example into account (which means that that example must be written down).

That's true even if the `RateLimitFilter` resource is meant to be an Envoy-independent piece of the Gateway API, because it doesn't make any sense for us to design it in a way that we can't actually use it.

So, overall: descriptors could indeed be an implementation detail! but the work of sorting out how the initial implementation can be done with code that already exists is important, since otherwise we're very likely to end up with something that the details prevent from actually being implemented. 🙂 

*: Obviously, we could also write a different Envoy filter to do rate limiting for Envoy Gateway. But let's not do that. 🙂

Reply:
thanks for explaining that @kflynn . In the interest of making finishing smaller work items faster, I prefer this document to be just the user facing API design, will add more design details on the mappings to descriptors in later PRs

Reply:
@arkodg So maybe the simplest way forward is to write out the multiple-overlapping-ratelimits case using `HTTPRoute`s and `RateLimitFilter`s, and then write a minimal "as an example, here's how this might configure the existing Envoy ratelimiter" -- sound reasonable?

Reply:
@kflynn ive added https://github.com/envoyproxy/gateway/pull/706/commits/e30e7f749e130368853e4520556347c3f9f7ede7 to address explaining overlapping limits
regarding descriptors i have a draft PR up here https://github.com/envoyproxy/gateway/pull/726 which implements API to descriptor translation, its incomplete since it depends on the API defined in this PR. Im hoping to keep any Envoy details out of this PR, make smaller incremental progress such as using this PR to design a user facing API and address all Envoy specifics in follow up PRs